### DISCUSSION TITLE
Error relation "oc_appconfig" does not exist

### DISCUSSION THREAD

**Discussion by froman753** [nextcloud/all-in-one#1553] [Questions] [answered]
I have followed the migration steps for migrating the files and database to the new AIO set up (AIO beta 4.0.0, both on 25.0.2). 
I copied over the old data directory, and modified the new config.php with the three values listed in the guide. All the containers are up and running, however, the Nextcloud container is running into an error. I actually restored from the initial AIO backup and ran through the steps again and still run into this same issue.

```
Installing imagemagick via apk...
Installing PHP build dependencies...
Installing Imagick via PECL...
Configuring Redis as session handler...
Setting php max children...
An unhandled exception has been thrown:
PDOException: SQLSTATE[42P01]: Undefined table: 7 ERROR:  relation "oc_appconfig" does not exist
LINE 1: SELECT * FROM "oc_appconfig"
                      ^ in /var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Connection.php:82

ace:
#0 /var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Connection.php(82): PDO->query('SELECT * FROM "...')
#1 /var/www/html/3rdparty/doctrine/dbal/src/Connection.php(1062): Doctrine\DBAL\Driver\PDO\Connection->query('SELECT * FROM "...')
#2 /var/www/html/lib/private/DB/Connection.php(261): Doctrine\DBAL\Connection->executeQuery('SELECT * FROM "...', Array, Array, NULL)
#3 /var/www/html/3rdparty/doctrine/dbal/src/Query/QueryBuilder.php(345): OC\DB\Connection->executeQuery('SELECT * FROM "...', Array, Array)
#4 /var/www/html/lib/private/DB/QueryBuilder/QueryBuilder.php(281): Doctrine\DBAL\Query\QueryBuilder->execute()
#5 /var/www/html/lib/private/AppConfig.php(415): OC\DB\QueryBuilder\QueryBuilder->execute()
#6 /var/www/html/lib/private/AppConfig.php(180): OC\AppConfig->loadConfigValues()
#7 /var/www/html/lib/private/AppConfig.php(371): OC\AppConfig->getApps()
#8 /var/www/html/lib/private/legacy/OC_App.php(965): OC\AppConfig->getValues(false, 'installed_versi...')
#9 /var/www/html/lib/private/Server.php(725): OC_App::getAppVersions()
#10 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(162): OC\Server->OC\{closure}(Object(OC\Server))
#11 /var/www/html/3rdparty/pimple/pimple/src/Pimple/Container.php(122): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(Pimple\Container))
#12 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(129): Pimple\Container->offsetGet('OC\\Memcache\\Fac...')
#13 /var/www/html/lib/private/ServerContainer.php(136): OC\AppFramework\Utility\SimpleContainer->query('OC\\Memcache\\Fac...', false)
#14 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(57): OC\ServerContainer->query('OC\\Memcache\\Fac...')
#15 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(184): OC\AppFramework\Utility\SimpleContainer->get('OC\\Memcache\\Fac...')
#16 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(162): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(OC\Server))
#17 /var/www/html/3rdparty/pimple/pimple/src/Pimple/Container.php(118): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(Pimple\Container))
#18 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(129): Pimple\Container->offsetGet('OCP\\ICacheFacto...')
#19 /var/www/html/lib/private/ServerContainer.php(136): OC\AppFramework\Utility\SimpleContainer->query('OCP\\ICacheFacto...', false)
#20 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(57): OC\ServerContainer->query('OCP\\ICacheFacto...')
#21 /var/www/html/lib/private/Server.php(1119): OC\AppFramework\Utility\SimpleContainer->get('OCP\\ICacheFacto...')
#22 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(162): OC\Server->OC\{closure}(Object(OC\Server))
#23 /var/www/html/3rdparty/pimple/pimple/src/Pimple/Container.php(122): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(Pimple\Container))
#24 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(129): Pimple\Container->offsetGet('OCP\\Lock\\ILocki...')
#25 /var/www/html/lib/private/ServerContainer.php(136): OC\AppFramework\Utility\SimpleContainer->query('OCP\\Lock\\ILocki...', false)
#26 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(57): OC\ServerContainer->query('OCP\\Lock\\ILocki...')
#27 /var/www/html/lib/private/Server.php(2070): OC\AppFramework\Utility\SimpleContainer->get('OCP\\Lock\\ILocki...')
#28 /var/www/html/lib/private/Files/View.php(117): OC\Server->getLockingProvider()
#29 /var/www/html/lib/private/Server.php(462): OC\Files\View->__construct()
#30 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(162): OC\Server->OC\{closure}(Object(OC\Server))
#31 /var/www/html/3rdparty/pimple/pimple/src/Pimple/Container.php(122): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(Pimple\Container))
#32 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(129): Pimple\Container->offsetGet('OC\\Files\\Node\\H...')
#33 /var/www/html/lib/private/ServerContainer.php(136): OC\AppFramework\Utility\SimpleContainer->query('OC\\Files\\Node\\H...', false)
#34 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(57): OC\ServerContainer->query('OC\\Files\\Node\\H...')
#35 /var/www/html/lib/private/Server.php(1472): OC\AppFramework\Utility\SimpleContainer->get('OC\\Files\\Node\\H...')
#36 /var/www/html/lib/base.php(617): OC\Server->boot()
#37 /var/www/html/lib/base.php(1144): OC::init()
#38 /var/www/html/console.php(48): require_once('/var/www/html/l...')
#39 /var/www/html/occ(11): require_once('/var/www/html/c...')
#40 {main}
```

Any advice for troubleshooting this? I may eventually give up and just try the user_migration app.

**Comment by szaimen** [comment]
Hi, can you post the database container logs here? It should autonatically import the database dump for you which should be visible in the logs.

**Reply by froman753** [reply]
Here is the database container logs: https://pastebin.com/WDpkEGZp

**Reply by szaimen** [reply]
Hm... It looks like the database dump was not imported... Did you really follow all steps correctly?

**Reply by szaimen** [reply]
I am especialy curious if you've applied all of these commands correctly:
![image](https://user-images.githubusercontent.com/42591237/208538997-0a50553f-d25c-4894-8cbe-4f67292fdde3.png)

**Reply by szaimen** [reply]
I am asking because I need the database container log from when it imports the database dump which should be visible in the container logs then.

**Reply by froman753** [reply]
Yes I did follow those steps as above. There was no errors, the commands just completed seemingly without issue. I will try to run them again with verbose. 
For modifying the database export such as in this step:
`Change it to look like this: local::/mnt/ncdata/.`

This is the same even if I had changed the standard path of your datadirectory, correct?

**Reply by froman753** [reply]
I reran those above commands and this is the database container log from the initial import
https://pastebin.com/SteWh7Yk

It does not seem to like the owner name of the database. Not sure how to fix it.

**Reply by szaimen** [reply]
It looks like we found the culprit then. Thanks for the logs! I'll have a look how to fix this going forward!

**Reply by froman753** [reply]
Is there no obvious fix for it? I'm eager to get back up and running

**Reply by szaimen** [reply]
The fix is in https://github.com/nextcloud/all-in-one/pull/1557/files and will get released with the next beta release which is currently planned for this friday.

**Comment by szaimen** [comment]
I've created https://github.com/nextcloud/all-in-one/pull/1557 which should make the database import easier to debug and cover this special case going forward.

**Comment by szaimen** [comment]
This is now fixed with v4.0.1 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

You will need to re-add your database-dump before updating your containers to 4.0.1 as otherwise you will need to recreate the database-dump since it assumes 25.0.2 as base otherwise.

**Reply by froman753** [reply]
The database seemed to import okay. The database container logs didn't have any errors until the nextcloud container started running.
Still running into the same issue with the "oc_appconfig" does not exist.

Database logs: https://pastebin.com/NwVbWTG4
Nextcloud logs: https://pastebin.com/9EdzqapH

**Reply by szaimen** [reply]
> Database logs: https://pastebin.com/NwVbWTG4

Hm no. The database logs do not look like they should. It should have imported your database dump on startup but it didnt do that. Did you make sure to put your database dump into the correct place before starting it again?

**Reply by froman753** [reply]
Database logs: https://pastebin.com/6CvHHk3X

It appears the import did fail when I retried it.

**Comment by szaimen** [comment]
> Database logs: https://pastebin.com/6CvHHk3X
> 
> It appears the import did fail when I retried it.

All.right! Thanks for testing! So it seems like my fix did not work. 

So unfortunately I don't see any good way to fix the problem on our side automatically. However I will try to add some documentation on this.

**Reply by froman753** [reply]
Thank you for the help. I appreciate the effort.

**Comment by szaimen** [comment]
I'll address the problem with https://github.com/nextcloud/all-in-one/pull/1620

**Comment by szaimen** [comment, marked as answer]
Hi, please see this, it is now part of the migration documentation: 
![image](https://user-images.githubusercontent.com/42591237/209954228-e194d27b-1d7e-4a87-9c24-73a71e77b549.png)

**Reply by szaimen** [reply]
The documentaiton was updated. It looks like this now: 
![image](https://github.com/nextcloud/all-in-one/assets/42591237/661bc6b1-4b6f-4649-8d3f-60f3ab857e86)

**Reply by dokterbob** [reply]
Link to the docs would facilitate cut-n-paste.

**Reply by szaimen** [reply]
https://github.com/nextcloud/all-in-one/blob/main/migration.md#migrate-the-files-and-the-database

**Comment by KenniVelez** [comment]
Thanks

**Comment by gbakeman** [comment]
Apologies for bumping this - I'm experiencing the exact same error as OP, and I did the migration myself to a PostgresSQL server and the standalone Nextcloud docker image. Originally the database was in a MySQL server, with a `nextcloud` db username. I'm exported this database into a PostgresQL server now, again with a user named `nextcloud`. In doing this manually, do I still need to worry about owner/usernames? What is the underlying cause of this error?

## Update
From what I'm understanding, this error indicates multiple possible deficiencies in my database that required fixing.

Since I had created the new nextcloud database under the root user (postgres) to be filled with my migrated data, the owner of that database was therefor the "root" user and not nextcloud's database user (the name doesn't appear to matter.) This meant that imported data was owned by the root user, and even though I added the nextcloud user with full permissions at the database level, these permissions did not propagate down into the schemas or even tables. In addition to that, initializing the database automatically created an empty, public [schema](https://www.postgresql.org/docs/current/ddl-schemas.html), which was created besides a nextcloud schema that contained my tables and data. Schemas act as these pseudo-containers for tables, and because the public schema was available to the nextcloud db user while the nextcloud **schema** was not, the user was accessing the public schema and finding it empty.

First step was deleting the unneeded public schema. After that, I made sure the nextcloud db user had full access at the schema level, then on all tables by running the command `GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA [nextcloud schema] TO [nextcloud db user];`. Doing these two things seems to have resolved the error for now. Hope this helps anyone else trying to migrate a PostgresQL database.

**Comment by skarpinis** [comment]
simple put half of things broken or not function as intended with this nextcloud.
